src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/TelemetrySession.cs

```diff
@@ -42,7 +45,53 @@ public static VsTelemetryEvent ToVsTelemetryEvent(TelemetryEvent telemetryEvent)
                 vsTelemetryEvent.Properties[VSPropertyNamePrefix + pair.Key] = new VsTelemetryPiiProperty(pair.Value);
             }
 
+            foreach (var pair in telemetryEvent.GetComplexData())
+            {
+                vsTelemetryEvent.Properties[VSPropertyNamePrefix + pair.Key] = ToComplexProperty(pair.Value);
+            }
+
             return vsTelemetryEvent;
         }
+
+        private static object ToComplexProperty(object value)
+        {
+            if (value is TelemetryEvent @event)
+            {
+                var dictionary = new Dictionary<string, object>();
+
+                foreach (var pair in @event)
+
+                {
+                    dictionary[pair.Key] = pair.Value;
+                }
+
+                foreach (var pair in @event.GetPiiData())
+                {
+                    dictionary[pair.Key] = new VsTelemetryPiiProperty(pair.Value);
+                }
+
+                foreach (var pair in @event.GetComplexData())
+                {
+                    dictionary[pair.Key] = ToComplexProperty(pair.Value);
+                }
+
+                return new VsTelemetryComplexProperty(dictionary);
+            }
+            else if (value is IEnumerable enumerable)
+            {
+                var list = new List<object>();
+
+                foreach (var item in list)
```

Comment:
I think there is a logical error here. foreach is iterating `list` which is empty because it is initialized to `new List<object>()` in the above line. Are we supposed to iterate `value` IEnumerable ?